// react-sharedb/lib/batching.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Batching = undefined;

var _set = require("babel-runtime/core-js/set");

var _set2 = _interopRequireDefault(_set);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Synchronous batching of functions execution
// Requires ES6 Set

var Batching = exports.Batching = function () {
  function Batching() {
    (0, _classCallCheck3.default)(this, Batching);

    this.active = false;
    this.queue = new _set2.default();
    this.flushActive = false;
  }

  (0, _createClass3.default)(Batching, [{
    key: "batch",
    value: function batch(fn) {
      if (this.active) return fn();
      this.active = true;
      fn();
      this.flush();
      this.active = false;
    }
  }, {
    key: "flush",
    value: function flush() {
      if (this.flushActive) return;
      this.flushActive = true;
      while (true) {
        if (this.queue.size === 0) break;
        var fn = getFirstItem(this.queue);
        this.queue.delete(fn);
        fn();
      }
      this.flushActive = false;
    }
  }, {
    key: "add",
    value: function add(fn) {
      if (!this.active) return fn();
      this.queue.add(fn);
    }
  }]);
  return Batching;
}();

function getFirstItem(set) {
  if (set.values) {
    var it = set.values();
    var first = it.next();
    return first.value;
    // Shim for IE
  } else {
    var _first = void 0;
    set.forEach(function (item) {
      if (_first) return;
      _first = item;
    });
    return _first;
  }
}

exports.default = new Batching();
