// react-sharedb/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.raw = exports.usePage = exports.useSession = exports.useQueryDoc = exports.useLocalDoc = exports.useQueryIds = exports.useNow = exports.useComponentId = exports.ComponentMetaContext = exports.useEmit = exports.useOn = exports.useModel = exports.emit = exports.useApi = exports.useValue = exports.useLocal = exports.useQuery = exports.useDoc = exports.destroyer = exports.observer = exports.subApi = exports.subValue = exports.subQuery = exports.subDoc = exports.subLocal = exports.clone = exports.initLocalCollection = exports._semaphore = exports.batchModel = exports.batch = exports.batching = exports.subscribe = exports.$root = exports.model = undefined;

var _model = require('@react-sharedb/model');

Object.defineProperty(exports, 'model', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_model).default;
  }
});
Object.defineProperty(exports, '$root', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_model).default;
  }
});

var _semaphore = require('./semaphore');

Object.defineProperty(exports, '_semaphore', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_semaphore).default;
  }
});

var _util = require('./util');

Object.defineProperty(exports, 'initLocalCollection', {
  enumerable: true,
  get: function get() {
    return _util.initLocalCollection;
  }
});
Object.defineProperty(exports, 'clone', {
  enumerable: true,
  get: function get() {
    return _util.clone;
  }
});

var _subscriptionTypeFns = require('./subscriptionTypeFns');

Object.defineProperty(exports, 'subLocal', {
  enumerable: true,
  get: function get() {
    return _subscriptionTypeFns.subLocal;
  }
});
Object.defineProperty(exports, 'subDoc', {
  enumerable: true,
  get: function get() {
    return _subscriptionTypeFns.subDoc;
  }
});
Object.defineProperty(exports, 'subQuery', {
  enumerable: true,
  get: function get() {
    return _subscriptionTypeFns.subQuery;
  }
});
Object.defineProperty(exports, 'subValue', {
  enumerable: true,
  get: function get() {
    return _subscriptionTypeFns.subValue;
  }
});
Object.defineProperty(exports, 'subApi', {
  enumerable: true,
  get: function get() {
    return _subscriptionTypeFns.subApi;
  }
});

var _observer = require('./hooks/observer');

Object.defineProperty(exports, 'observer', {
  enumerable: true,
  get: function get() {
    return _observer.observer;
  }
});

var _destroyer = require('./hooks/destroyer');

Object.defineProperty(exports, 'destroyer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_destroyer).default;
  }
});

var _types = require('./hooks/types');

Object.defineProperty(exports, 'useDoc', {
  enumerable: true,
  get: function get() {
    return _types.useDoc;
  }
});
Object.defineProperty(exports, 'useQuery', {
  enumerable: true,
  get: function get() {
    return _types.useQuery;
  }
});
Object.defineProperty(exports, 'useLocal', {
  enumerable: true,
  get: function get() {
    return _types.useLocal;
  }
});
Object.defineProperty(exports, 'useValue', {
  enumerable: true,
  get: function get() {
    return _types.useValue;
  }
});
Object.defineProperty(exports, 'useApi', {
  enumerable: true,
  get: function get() {
    return _types.useApi;
  }
});

var _helpers = require('./hooks/helpers');

Object.defineProperty(exports, 'emit', {
  enumerable: true,
  get: function get() {
    return _helpers.emit;
  }
});
Object.defineProperty(exports, 'useModel', {
  enumerable: true,
  get: function get() {
    return _helpers.useModel;
  }
});
Object.defineProperty(exports, 'useOn', {
  enumerable: true,
  get: function get() {
    return _helpers.useOn;
  }
});
Object.defineProperty(exports, 'useEmit', {
  enumerable: true,
  get: function get() {
    return _helpers.useEmit;
  }
});
Object.defineProperty(exports, 'ComponentMetaContext', {
  enumerable: true,
  get: function get() {
    return _helpers.ComponentMetaContext;
  }
});
Object.defineProperty(exports, 'useComponentId', {
  enumerable: true,
  get: function get() {
    return _helpers.useComponentId;
  }
});
Object.defineProperty(exports, 'useNow', {
  enumerable: true,
  get: function get() {
    return _helpers.useNow;
  }
});
Object.defineProperty(exports, 'useQueryIds', {
  enumerable: true,
  get: function get() {
    return _helpers.useQueryIds;
  }
});
Object.defineProperty(exports, 'useLocalDoc', {
  enumerable: true,
  get: function get() {
    return _helpers.useLocalDoc;
  }
});
Object.defineProperty(exports, 'useQueryDoc', {
  enumerable: true,
  get: function get() {
    return _helpers.useQueryDoc;
  }
});
Object.defineProperty(exports, 'useSession', {
  enumerable: true,
  get: function get() {
    return _helpers.useSession;
  }
});
Object.defineProperty(exports, 'usePage', {
  enumerable: true,
  get: function get() {
    return _helpers.usePage;
  }
});

var _observerUtil = require('@nx-js/observer-util');

Object.defineProperty(exports, 'raw', {
  enumerable: true,
  get: function get() {
    return _observerUtil.raw;
  }
});

require('./globalInit');

var _subscribe = require('./subscribe');

var _subscribe2 = _interopRequireDefault(_subscribe);

var _batching = require('./batching');

var _batching2 = _interopRequireDefault(_batching);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.subscribe = _subscribe2.default;
exports.batching = _batching2.default;
var batch = exports.batch = _batching2.default.batch.bind(_batching2.default);
// TODO: DEPRECATED. Add warning to use batch instead
var batchModel = exports.batchModel = batch;
