// react-sharedb/lib/types/Base.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Base = function () {
  function Base(model, key, params) {
    (0, _classCallCheck3.default)(this, Base);

    this.key = key;
    this.params = params;
    this.model = model;
  }

  (0, _createClass3.default)(Base, [{
    key: "refModel",
    value: function refModel() {}
  }, {
    key: "unrefModel",
    value: function unrefModel() {}
  }, {
    key: "destroy",
    value: function destroy() {
      this.cancel();
      delete this.model;
      delete this.params;
      delete this.key;
    }

    // Cancel initialization process

  }, {
    key: "cancel",
    value: function cancel() {
      // If model doesn't exist, it means that the item was already destroyed,
      // so no need to cancel
      if (!this.cancelled) this.cancelled = true;
    }
  }]);
  return Base;
}();

exports.default = Base;
module.exports = exports['default'];
