// react-sharedb/lib/hooks/destroyer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require("babel-runtime/helpers/toConsumableArray");

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Destroyer = function () {
  function Destroyer() {
    (0, _classCallCheck3.default)(this, Destroyer);

    this.fns = [];
  }

  (0, _createClass3.default)(Destroyer, [{
    key: "add",
    value: function add(fn) {
      this.fns.push(fn);
    }
  }, {
    key: "run",
    value: function run() {
      this.fns.forEach(function (fn) {
        return fn();
      });
      this.fns.length = 0;
    }
  }, {
    key: "getDestructor",
    value: function getDestructor() {
      var fns = [].concat((0, _toConsumableArray3.default)(this.fns));
      this.fns.length = 0;
      return function () {
        fns.forEach(function (fn) {
          return fn();
        });
        fns.length = 0; // force remove function refs
      };
    }
  }, {
    key: "reset",
    value: function reset() {
      this.fns.length = 0;
    }
  }]);
  return Destroyer;
}();

exports.default = new Destroyer();
module.exports = exports['default'];
